etrago/tools/constraints.py

```diff
@@ -1183,6 +1194,211 @@ def _capacity_factor_per_gen_cntr_nmp(self, network, snapshots):
+    }
+
+    engine = db.connection(section=self.args["db"])
+    try:
+        sql = f"""
+        SELECT gas_parameters
+        FROM scenario.egon_scenario_parameters
+        WHERE name = '{scn_name}';"""
+        df = pd.read_sql(sql, engine)
+        arg = df["max_gas_generation_overtheyear"]
+    except:
+        arg = arg_def[scn_name]
+
+    return arg
+
+
+def add_ch4_constraints(self, network, snapshots):
+    """
+    Add CH4 constraints for optimization with pyomo
+
+    Functionality that limits the dispatch of CH4 generators. In
+    Germany, there is one limitation specific for biogas and one
+    limitation specific for natural gas (natural gas only in eGon2035).
+    Abroad, each generator has its own limitation contains in the
+    column e_nom_max.
+
+    Parameters
+    ----------
+    network : :class:`pypsa.Network
+        Overall container of PyPSA
+    snapshots : pandas.DatetimeIndex
+        List of timesteps considered in the optimization
+
+    Returns
+    -------
+    None.
+    """
+    scn_name = self.args["scn_name"]
+    n_snapshots = self.args["end_snapshot"] - self.args["start_snapshot"] + 1
+
+    # Add constraint for Germany
+    arg = read_max_gas_generation(self)
+    gas_carrier = arg.keys()
+
+    carrier_names = {
+        "eGon2035": {"CH4": "CH4_NG", "biogas": "CH4_biogas"},
+        "eGon100RE": {"biogas": "CH4"},
+    }
+
+    for c in gas_carrier:
+        gens = network.generators.index[
+            (network.generators.carrier == carrier_names[scn_name][c])
+            & (
+                network.generators.bus.astype(str).isin(
+                    network.buses.index[network.buses.country == "DE"]
+                )
+            )
+        ]
+        if not gens.empty:
+            factor = arg[c]
+
+            def _rule_max(m):
+
+                dispatch = sum(
+                    m.generator_p[gen, sn] * network.snapshot_weightings.generators[sn]
+                    for gen in gens
+                    for sn in snapshots
+                )
+
+                return dispatch <= factor * (n_snapshots / 8760)
+
+            setattr(network.model, "max_flh_DE_" + c, Constraint(rule=_rule_max))
+
+    # Add contraints for neigbouring countries
+    gen_abroad = network.generators[
+        (network.generators.carrier == "CH4")
+        & (
+            network.generators.bus.astype(str).isin(
+                network.buses.index[network.buses.country != "DE"]
+            )
+        )
+        & (network.generators.e_nom_max != np.inf)
+    ]
+    for g in gen_abroad.index:
+        factor = network.generators.e_nom_max[g]
+
+        def _rule_max(m):
+
+            dispatch = sum(
+                m.generator_p[g, sn] * network.snapshot_weightings.generators[sn]
+                for sn in snapshots
+            )
+
+            return dispatch <= factor * (n_snapshots / 8760)
+
+        setattr(network.model, "max_flh_abroad_" + str(g), Constraint(rule=_rule_max))
```

Comment:
The indices stored in `g` include spaces, so the names of the constraints also include some spaces. This makes it a bit more complicated to access the constraints when you want to have a look at them. Could you maybe replace the space with a `_` in the names of the constraints? 
(should be only: `g.replace(' ', '_') `)